### ISSUE TITLE
Struct constructor doesn't require initialization of class reference

### ISSUE THREAD

**Issue by ntrel** [Neat-Lang/neat#21]
```d
module objembed;

class C
{
    int i;
}

struct S
{
    C c;
    this() {}
}

void main()
{
    //S s; // error
    S s = S();
}
```
Declaring `S s;` gives:
```
      ~ Cannot declare s without initializer: S is not zero-initializable
```
But declaring `S.this` without initialization of `c` is not caught. `s.c` is then `null`.

**Comment by FeepingCreature** [comment]
Yes, this is ... kind of hard to fix. The problem is that we'd need control flow analysis (or runtime errors) to catch cases where there exist flows through the constructor that don't set a field. At the limit, this is no-shit halting-problem tier undecidable.

**Comment by ntrel** [comment]
Yes, some kind of basic flow analysis would be needed. If implemented like cppfront, the following code would error:
```d
this(){
    if (random!bool()) // line 12
        c = new C; 
}
```
```
objembed:12: Error: `if` statement initializes `c` in only one branch
```
Requiring `if` to initialize a member in both branches if it is initialized in either seems a pragmatic way to solve the halting problem.

**Comment by FeepingCreature** [comment]
Okay, I'm gonna say that yes, I could do that, but I'm not gonna in the short-term. It seems a lot of effort and a lot of compiler machinery that wouldn't be used anywhere else.

Patches, as usual, welcome!

**Comment by ntrel** [comment]
> It seems a lot of effort for a mid improvement

For now perhaps an out contract for structs with non-nullable members could be added in constructors that asserts that those members are not null. Later the `if` branch checking could be implemented and the out contracts removed (though the former can have false positives unlike the latter).

> a lot of compiler machinery that wouldn't be used anywhere else

This would also be useful to ensure a struct with both a defined constructor and destructor is actually constructed and not left as T.init. This would guarantee that such structs do not have to handle T.init in their destructor (which has a runtime cost as well as a design cost to detect the init value). Then if there is no initialization of the struct, the compiler can avoid calling the destructor or allowing it to be copied. (I also raised this [here](https://github.com/Neat-Lang/neat/discussions/19#discussioncomment-4571135)).

> Patches, as usual, welcome!

Thanks, I haven't looked at the source yet. (I'm a bit put off by the compile time as I'm spoilt by dmd ;-)).

**Comment by FeepingCreature** [comment]
Initial compile time is high, but follow-up should be fine.

`./build.sh` builds the compiler with optimization. This is usually the right move, but can be turned off with `FAST= ./build.sh`.

That's a good idea with the out-condition assert!
